smenon/edit_multiple_shapes

sidmenon97 opened the pull request:
- Changes made to VitruvioComponentDetails.cpp/h
- If the RPKs for the selected objects are different, it acts as before
- Otherwise, it now allows editing the multiple selected objects simultaneously
- Added a function to get all Vitruvio Components that are selected
- Added a function to determine if the selected Vitruvio Components have different values in its field
- Modified widget functions/Added new widget functions to account for multiple values selected

sidmenon97 committed:
Can edit multiple shapes simultaneously

Added capability to edit multiple Vitruvio Components simultaneously

Fix tabs

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -96,6 +98,23 @@ bool IsVitruvioComponentSelected(const TArray<TWeakObjectPtr<UObject>>& ObjectsB
 	return false;
 }
 
+void GetAllVitruvioComponents(const TArray<TWeakObjectPtr<UObject>>& ObjectsBeingCustomized, 
```
I think `IsVitruvioComponentSelected` is no longer needed after this change and could be deleted.

bneukom replied:
Consider returing the ObjectsBeingCustomized by value from the function instead of passing by reference (would also make the callsite a bit easier to understand).

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -96,6 +98,23 @@ bool IsVitruvioComponentSelected(const TArray<TWeakObjectPtr<UObject>>& ObjectsB
+void GetAllVitruvioComponents(const TArray<TWeakObjectPtr<UObject>>& ObjectsBeingCustomized, 
+								TArray<UVitruvioComponent*>& OutVitruvioComponents)
+{
+	for (size_t ObjectIndex = 0; ObjectIndex < ObjectsBeingCustomized.Num(); ++ObjectIndex)
```
We usually prefer the range based foor loop instead of accessing the elements via their index (if it is not needed)

Eg: `for (const TWeakObjectPtr<UObject>& CurrentObject : ObjectsBeingCustomized)`

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -243,6 +262,87 @@ TSharedPtr<SHorizontalBox> CreateTextInputWidget(TSharedPtr<IPropertyHandle> Str
 }
 
+template <typename Attr>
+TSharedPtr<SHorizontalBox> CreateMultipleValueFloatAttributeWidget(Attr* Attribute, TSharedPtr<IPropertyHandle> Property)
```
The `Property` could be passed by const reference.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -243,6 +262,87 @@ TSharedPtr<SHorizontalBox> CreateTextInputWidget(TSharedPtr<IPropertyHandle> Str
+	// clang-format on
+}
+
+TSharedPtr<SHorizontalBox> CreateMultipleValueStringAttributeWidget(TSharedPtr<IPropertyHandle> Property)
```
Seme here by const reference.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -459,8 +573,25 @@ void AddScalarWidget(const TArray<TSharedRef<IDetailTreeNode>> DetailTreeNodes,
+bool DifferentValuesSelected(URuleAttribute* Attribute, FString AttributeKey, TArray<UVitruvioComponent*> VitruvioComponents)
+{
+	FString AttributeValue = Attribute->GetValueAsString();
+	for (size_t Component = 0; Component < VitruvioComponents.Num(); ++Component)
```
Range based for loop instead.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -459,8 +573,25 @@ void AddScalarWidget(const TArray<TSharedRef<IDetailTreeNode>> DetailTreeNodes,
 	}
 }
 
+bool DifferentValuesSelected(URuleAttribute* Attribute, FString AttributeKey, TArray<UVitruvioComponent*> VitruvioComponents)
```
Parameters could be passed by const reference.

bneukom replied:
Also consider renaming the function to something more descriptive (the selected postfix is only "implied" by the callsite).

Maybe something like `AreValuesDifferent` (or something like that, up to you).

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -760,30 +892,46 @@ void FVitruvioComponentDetails::BuildAttributeEditor(IDetailLayoutBuilder& Detai
+				// Issue: Undo doesn't undo the change in attributes on all vitruvio actors
+				TMap<FString, FString> ChangedAttributes;
+				ChangedAttributes.Add(AttributeKey, Attribute->GetValueAsString());
+				for (size_t i = 0; i < VitruvioComponentsSelected.Num(); ++i)
```
Range based for loop.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -760,30 +892,46 @@ void FVitruvioComponentDetails::BuildAttributeEditor(IDetailLayoutBuilder& Detai
+				ChangedAttributes.Add(AttributeKey, Attribute->GetValueAsString());
+				for (size_t i = 0; i < VitruvioComponentsSelected.Num(); ++i)
+				{
+					VitruvioComponentsSelected[i]->SetAttributes(ChangedAttributes);
+					VitruvioComponentsSelected[i]->EvaluateRuleAttributes(VitruvioComponentsSelected[i]->GenerateAutomatically);
```
It could make sense to introduce a generic `SetAttribute` function on the `VitruvioComponent` (which takes a two Strings key and value) (internally the SetAttributes could then use this SetAttribute function to reduce code duplication).

Then we wouldn't have to abuse the `SetAttributes` function?

Not necessary to change this. What do you think?

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -760,30 +892,46 @@ void FVitruvioComponentDetails::BuildAttributeEditor(IDetailLayoutBuilder& Detai
+				ChangedAttributes.Add(AttributeKey, Attribute->GetValueAsString());
+				for (size_t i = 0; i < VitruvioComponentsSelected.Num(); ++i)
+				{
+					VitruvioComponentsSelected[i]->SetAttributes(ChangedAttributes);
```
Also here we apply the change twice to the original `VitruvioActor` variable I think? I think an `if (VitruvioComponentSelected != VitruvioActor->GetVitruvioComponent)` would fix that.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -760,30 +892,46 @@ void FVitruvioComponentDetails::BuildAttributeEditor(IDetailLayoutBuilder& Detai
 		Generators.Add(Generator);
 
+		bool MultipleValuesSelected = false;
+		if (VitruvioComponentsSelected.Num() > 1)
+		{
```
Consider adding this code to the function below so the callsite would only look like this:

`bool MultipleValuesSelected = DifferentValuesSelected(....`

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioComponentDetails.cpp:
```diff
@@ -856,26 +1004,39 @@ void FVitruvioComponentDetails::CustomizeDetails(IDetailLayoutBuilder& DetailBui
-		return;
+
+		URulePackage* ComponentRPK = VitruvioComponentsSelected[0]->GetRpk();
+		for (size_t ComponentIndex = 0; ComponentIndex < VitruvioComponentsSelected.Num(); ++ComponentIndex)
```
Range based foor loop.

bneukom commented on VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Public/VitruvioComponentDetails.h:
```diff
@@ -45,6 +45,7 @@ class FVitruvioComponentDetails final : public IDetailCustomization
 	void BuildAttributeEditor(IDetailLayoutBuilder& DetailBuilder, IDetailCategoryBuilder& RootCategory, UVitruvioComponent* VitruvioActor);
 
 	TArray<TWeakObjectPtr<UObject>> ObjectsBeingCustomized;
+	TArray<UVitruvioComponent*> VitruvioComponentsSelected;
```
Consider reanming this to `SelectedVitruvioComponents`

bneukom reviewed (approved):
I rebased the PR branch to the latest changes on develop (and force pushed). Please reset your branch before making any changes :)

Very very nice work! I like how you adhered to all Unreal guidelines so well already 👍 I also did a quick test and it worked really well (will have to do a bit more indepth functional testing though after we merged it).

The comments are all relatively minor and some code style things.

Also please squash your second commit with the first one (the formatting changes).

sidmenon97 committed:
Updated to range-based for loops and renaming functions

Added generic SetAttribute function to VitruvioComponent

* Added generic SetAttribute function to VitruvioComponent and updated SetAttributes() to use the new SetAttrbiute function.
* Also, some additional variable renaming changes.

bneukom reviewed (changes requested):
Some small code change requests.

Overall I think the `bool` (checkbox) input fields as well as the `enum` (dropdown) widgets could also have a different appearance when there are multiple values selected?

In Unreal it looks like this:

![image](https://github.com/user-attachments/assets/ce3a65e9-8aba-4251-bffd-3b49162b5a7a)

Is that something we could add as well?

I also rebased the branch to the latest develop again (and force pushed). I think you merged the main branch into this branch at some point which made the commit histry a bit strange.

sidmenon97 committed:
Bool and Enum widgets modified

- bool and enum widgets are now modified to also show 'Multiple Values' behavior.
- MultipleValueStringAttributeWidget merged with the original Text Input widget.
